### PR TITLE
FIX: Minor fixes to GL tutorial examples

### LINKED ISSUE #576: timer.start() causes gap in timer callbacks
With Qt (and possibly other backends), calling `timer.start()` when the timer is already running causes the timer to cancel its next timeout and schedule a new one. This results in a gap in the timer callbacks, and can cause choppy behavior if, for example, the user calls `timer.start()` whenever a mouse event is handled.

### LINKED ISSUE #630: GLFW Backend Crash when Shutting Down on OSX
With Python3.4 os OSX, I get a crash on shutdown when telling VisPy to use the GLFW backend.  I've tried to reproduce it with lighter weight applications like the scene demos, but so far those seem OK.

The crash happens in _glfw.py and is caused by the event processing:

````
diff --git a/vispy/app/backends/_glfw.py b/vispy/app/backends/_glfw.py
index 9aeaf95..0c729ab 100644
--- a/vispy/app/backends/_glfw.py
+++ b/vispy/app/backends/_glfw.py
@@ -341,7 +342,9 @@ class CanvasBackend(BaseCanvasBackend):
             # glfw.glfwSetWindowShouldClose()  # Does not really cause a close
             self._vispy_set_visible(False)
             self._id, id_ = None, self._id
-            glfw.glfwPollEvents()
+
+            # [alexjc] Calling this leads to a crash.
+            # glfw.glfwPollEvents()
             glfw.glfwDestroyWindow(id_)
             gc.collect()  # help ensure context gets destroyed
````

Commenting out the poll for events has no other side-effects but fixing the crash in my application, but I'm curious about the original intent and whether this is a problem on other platforms.

### LINKED ISSUE #667: Event Ordering and Synchronous Processing Bugs (OSX)
I'm going through the test failures on OSX with Pyglet and GLFW, and there are about a dozen.  (With mixed success, see #666.)

One of the biggest categories of bugs is related to event processing.  I've seen two in particular:

**1. Synchronous Processing**

In some demos, when setting `self.size` at the top of the constructor, then on_resize is processed immediately and fails because the rest of the constructor has not finished processing.  (This was most common with GLFW.)

This is the case for `examples/basic/gloo/hello_fbo.py` in particular:

````
  File "examples/basics/gloo/hello_fbo.py", line 79, in __init__
    self.size = 560, 420
  File "vispy/app/canvas.py", line 296, in size
    return self._backend._vispy_set_size(size[0], size[1])
  File "vispy/app/backends/_glfw.py", line 319, in _vispy_set_size
    glfw.glfwSetWindowSize(self._id, w, h)
  File "vispy/app/backends/_glfw.py", line 390, in _on_draw
    self._vispy_canvas.events.draw(region=None)  # (0, 0, w, h))
  File "vispy/util/event.py", line 418, in __call__
    self._invoke_callback(cb, event)
  File "vispy/util/event.py", line 439, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "vispy/util/event.py", line 435, in _invoke_callback
    cb(event)
  File "examples/basics/gloo/hello_fbo.py", line 105, in on_draw
    with self._fbo:
AttributeError: 'Canvas' object has no attribute '_fbo'
````

There are two fixes for this, one is to assume that events are processed synchronously and declare this a bug in the demo.  The second is to make all events asynchronous, and only process `on_size` once the constructor has finished.

**2. Event Ordering**

Similarly, in Pyglet the events are not processed synchronously but it seems they are done in the wrong order.  For example in `examples/tutorial/gloo/textured_cube.py`:

````
  File "vispy/app/canvas.py", line 421, in __enter__
    self._backend._vispy_warmup()
  File "vispy/app/backends/_pyglet.py", line 220, in _vispy_warmup
    self._vispy_canvas.app.process_events()
  File "vispy/app/application.py", line 77, in process_events
    return self._backend._vispy_process_events()
  File "vispy/app/backends/_pyglet.py", line 154, in _vispy_process_events
    window.dispatch_events()
  File "pyglet/_build/lib/pyglet/window/cocoa/__init__.py", line 294, in dispatch_events
    self.dispatch_pending_events()
  File "pyglet/_build/lib/pyglet/window/cocoa/__init__.py", line 324, in dispatch_pending_events
    EventDispatcher.dispatch_event(self, *event)
  File "pyglet/_build/lib/pyglet/event.py", line 365, in dispatch_event
    if getattr(self, event_type)(*args):
  File "vispy/app/backends/_pyglet.py", line 315, in on_resize
    self._vispy_canvas.events.resize(size=(w, h))
  File "vispy/util/event.py", line 418, in __call__
    self._invoke_callback(cb, event)
  File "vispy/util/event.py", line 439, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "vispy/util/event.py", line 435, in _invoke_callback
    cb(event)
  File "examples/tutorial/gloo/textured_cube.py", line 91, in on_resize
    self.program['projection'] = projection
AttributeError: 'Canvas' object has no attribute 'program'
````

I've not traced through this specific example (there are many others failing too), but it can only happen if `on_size` is called before `on_initialize`.  This one I think must be fixed in the Pyglet backend.

I don't know why none of these bugs are visible on Linux or Windows, but they're easy to reproduce on OSX.  I'll try tackling the ones with GLFW incrementally, as I have been doing so far, but any advice/help is welcome!

### LINKED ISSUE #806: Fix GL examples on glfw backend (OSX at least)
Some of these examples have both `__init__` and `on_initialize` methods. They're not executed at the same time, and the examples look different or fail to work entirely.

```
examples/demo/gloo/galaxy/galaxy.py
examples/tutorial/gl/fireworks.py
examples/tutorial/gl/cube.py
```

Example of behaviour:
```
demo/gloo/galaxy$ python galaxy.py --vispy-backend=qt
begin init
end init
begin on_init
end on_init

demo/gloo/galaxy$ python galaxy.py --vispy-backend=glfw
INFO: vispy backend: glfw
begin init
begin on_init
end on_init
end init
```

### LINKED ISSUE #837: examples/tutorial/gl/fireworks.py doesn't work on OS X
Shows a black screen on any backend...

### LINKED ISSUE #838: Bunch of tests fail on Mac OS X
Mac OS X 10.10.1

Retina MacBook Pro 15" (mid-2012)

nVidia GeForce GT 650M

```
______________________________________________________________________ test_functionality_pyopengl ______________________________________________________________________

args = (), kwargs = {}

    def skipper_func(*args, **kwargs):
        """Skipper for normal test functions."""
        if skip_val():
            raise nose.SkipTest(get_msg(f, msg))
        else:
>           return f(*args, **kwargs)

../../../miniconda/envs/phy/lib/python3.4/site-packages/numpy/testing/decorators.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../miniconda/envs/phy/lib/python3.4/site-packages/numpy/testing/decorators.py:146: in skipper_func
    return f(*args, **kwargs)
vispy/gloo/gl/tests/test_functionality.py:53: in test_functionality_pyopengl
    _test_functonality('pyopengl2')
vispy/gloo/gl/tests/test_functionality.py:87: in _test_functonality
    objects = _prepare_vis()
vispy/gloo/gl/tests/test_functionality.py:226: in _prepare_vis
    assert_equal(gl.glGetShaderInfoLog(hvert), '')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

shader = 2

    def glGetShaderInfoLog(shader):
        res = GL.glGetShaderInfoLog(shader)
>       return res.decode('utf-8')
E       AttributeError: 'str' object has no attribute 'decode'

vispy/gloo/gl/_pyopengl2.py:132: AttributeError
_____________________________________________________
```

```
________________________________________________________________________ test_multiple_canvases _________________________________________________________________________

args = (), kwargs = {}

    def skipper_func(*args, **kwargs):
        """Skipper for normal test functions."""
        if skip_val():
            raise nose.SkipTest(get_msg(f, msg))
        else:
>           return f(*args, **kwargs)

../../../miniconda/envs/phy/lib/python3.4/site-packages/numpy/testing/decorators.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @requires_application()
    def test_multiple_canvases():
        """Testing multiple canvases"""
        n_check = 3
        app = use_app()
        with Canvas(app=app, size=_win_size, title='same_0') as c0:
            with Canvas(app=app, size=_win_size, title='same_1') as c1:
                ct = [0, 0]
    
                @c0.events.draw.connect
                def draw0(event):
                    ct[0] += 1
                    c0.update()
    
                @c1.events.draw.connect  # noqa, analysis:ignore
                def draw1(event):
                    ct[1] += 1
                    c1.update()
    
                c0.show()  # ensure visible
                c1.show()
                c0.update()  # force first draw
                c1.update()
    
                timeout = time() + 2.0
                while (ct[0] < n_check or ct[1] < n_check) and time() < timeout:
                    app.process_events()
                print((ct, n_check))
>               assert n_check <= ct[0] <= n_check + 1
E               assert 5 <= (3 + 1)

vispy/app/tests/test_simultaneous.py:70: AssertionError
------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------
([5, 4], 3)
```
```
______________________________________________________________________ test_functionality_pyopengl ______________________________________________________________________

args = (), kwargs = {}

    def skipper_func(*args, **kwargs):
        """Skipper for normal test functions."""
        if skip_val():
            raise nose.SkipTest(get_msg(f, msg))
        else:
>           return f(*args, **kwargs)

../../../miniconda/envs/phy/lib/python3.4/site-packages/numpy/testing/decorators.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../miniconda/envs/phy/lib/python3.4/site-packages/numpy/testing/decorators.py:146: in skipper_func
    return f(*args, **kwargs)
vispy/gloo/gl/tests/test_functionality.py:53: in test_functionality_pyopengl
    _test_functonality('pyopengl2')
vispy/gloo/gl/tests/test_functionality.py:87: in _test_functonality
    objects = _prepare_vis()
vispy/gloo/gl/tests/test_functionality.py:226: in _prepare_vis
    assert_equal(gl.glGetShaderInfoLog(hvert), '')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

shader = 2

    def glGetShaderInfoLog(shader):
        res = GL.glGetShaderInfoLog(shader)
>       return res.decode('utf-8')
E       AttributeError: 'str' object has no attribute 'decode'

vispy/gloo/gl/_pyopengl2.py:132: AttributeError
```

### PR DESCRIPTION
**Pull request by Eric89GXL**
@nippoo can you see if the examples you mention in #806 work with this PR?

Closes #838.
Closes #837.
Closes #806.
Closes #667.
Closes #630.
Closes #576.

### CODE REVIEW

**Comment by nippoo** [comment]
Galaxy and cube work great now! Fireworks shows a black screen on both qt4/glfw backends, though this has always been the case - never worked for me. I'll try debugging it in a bit. (Aren't you supposed to be enjoying your holiday?!)

**Comment by Eric89GXL** [comment]
Good to hear, thanks. Who says this isn't enjoyable? :)

**Comment by Eric89GXL** [comment]
@nippoo could you also run the test suite and see if it passes? Then maybe we can close #667 as well.

**Comment by Eric89GXL** [comment]
FWIW tests pass on OSX over here on pyqt4 and glfw.

**Comment by nippoo** [comment]
@Eric89GXL all tests pass apart from the 3 that normally fail, and the GLFW correlation matrix errors, so all good from here.

@rougier or @izaid: can you check if examples/tutorial/gl/fireworks.py works for you? This is just a black canvas for me on any backend, so maybe it's some unsupported GL call or something (but shouldn't be...)

**Comment by Eric89GXL** [comment]
Which 3 normally fail for you?

Looks like the GL fireworks doesn't work for me on GLFW or Pyqt4 on OSX, either, actually :(

**Comment by izaid** [comment]
On Mac, fireworks.py fails for me. Just a black screen.

**Comment by nippoo** [comment]
Okay I've submitted #837 for the black screen issue, #838 for the tests failing...

**Comment by Eric89GXL** [comment]
Thanks for opening the issues, @nippoo can you test the latest version of this PR to see if it fixes test failures (#838)?

**Comment by Eric89GXL** [comment]
I also pushed a small fix that should make Travis have fewer download failures, since we've had more of those lately.

**Comment by Eric89GXL** [comment]
@nippoo try again when you get a chance, pushed another fix for that error. (Sorry to be a pain, but at least you'll have passing tests soon!)

**Comment by nippoo** [comment]
Fixed for me, only a ton of correlation matrix errors now...

**Comment by Eric89GXL** [comment]
@almarklein feel free to review and merge this if you're happy.

File: examples/tutorial/gl/cube.py [hunk 1]
```diff
@@ -210,8 +209,10 @@ def on_draw(self, event):
                           gl.GL_UNSIGNED_INT, None)
 
     def on_resize(self, event):
-        width, height = event.size
-        gl.glViewport(0, 0, width, height)
+        self._resize(*(event.size + event.physical_size))
+
+    def _resize(self, width, height, physical_width, physical_height):
+        gl.glViewport(0, 0, physical_width, physical_height)
         projection = perspective(35.0, width / float(height), 2.0, 10.0)
```
Comment (almarklein) [inline examples/tutorial/gl/cube.py]: I did not follow the discussions regarding to `on_initialize` and `on_resize` that closely. But should it not be that the backend calls `on_resize` before the first draw call?

Comment (Eric89GXL) [inline examples/tutorial/gl/cube.py]: It is not guaranteed by every backend.

Comment (almarklein) [inline examples/tutorial/gl/cube.py]: Should we make it so? I've also relied on it subconsciously.

**Comment by almarklein** [comment]
Other than `on_resize` issue, LGTM

File: examples/tutorial/gl/cube.py (hunk 1 above)
Comment (Eric89GXL) [inline examples/tutorial/gl/cube.py]: Probably at some point, yeah. I'll open an issue so it doesn't hold this one up :)

**Comment by Eric89GXL** [comment]
@almarklein @nippo is happy with this, I rebased and opened #841 about the resizing issue -- feel free to merge if Travis comes back happy.

**Pull request merged**